File: sdk/typescript/packages/api/src/sampling.ts
import { UnimplementedHook } from './hooks';
import type { HookContext, HookHints, EventEnvelope } from './hooks';

export type SamplingStrategy = 'user_id_hash' | 'random' | 'none';

export interface SamplingPolicy {
  strategy: SamplingStrategy;
  rate: number;
}

export type SamplingLookup = (eventName: string) => SamplingPolicy | undefined;

// SampledError is thrown by SamplingHook.before when an event is dropped by the sampling policy.
export class SampledError extends Error {
  constructor() {
    super('event dropped by sampling policy');
    this.name = 'SampledError';
  }
}

// SamplingHook applies the sampling policy declared for an event during the Before stage.
// Sampled-out events throw SampledError; events without a policy pass through unchanged.
export class SamplingHook extends UnimplementedHook {
  constructor(private readonly lookup: SamplingLookup) {
    super();
  }

  override async before(hc: HookContext, _hints?: HookHints): Promise<EventEnvelope | null> {
    const policy = this.lookup(hc.eventName);
    if (!policy || !policy.strategy || policy.strategy === 'none') return null;

    let keep: boolean;
    switch (policy.strategy) {
      case 'user_id_hash':
        keep = hashKeep(hc.context.userId ?? '', policy.rate);
        break;
      case 'random':
        keep = Math.random() < policy.rate;
        break;
      default:
        return null;
    }

    if (!keep) throw new SampledError();
    return null;
  }
}

// fnv1a32 computes the FNV-1a 32-bit hash of s, operating on UTF-8 bytes.
// Matches Go's hash/fnv New32a implementation exactly.
function fnv1a32(s: string): number {
  const bytes = new TextEncoder().encode(s);
  let h = 2166136261; // FNV offset basis (32-bit)
  for (const byte of bytes) {
    h ^= byte;
    h = Math.imul(h, 16777619) >>> 0; // FNV prime, keep 32-bit unsigned
  }
  return h >>> 0;
}

// hashKeep returns true when userId hashes into the kept fraction [0, rate).
// The decision is deterministic: the same userId always produces the same outcome.
function hashKeep(userId: string, rate: number): boolean {
  const norm = fnv1a32(userId) / 0x100000000; // normalise to [0, 1)
  return norm < rate;
}

File: sdk/kotlin/api/src/main/kotlin/io/eventspec/analytics/Sampling.kt
package io.eventspec.analytics

enum class SamplingStrategy {
  USER_ID_HASH,
  RANDOM,
  NONE
}

data class SamplingPolicy(
    val strategy: SamplingStrategy,
    val rate: Double,
)

class SampledError : Exception("event dropped by sampling policy")

typealias SamplingLookup = (eventName: String) -> SamplingPolicy?

// SamplingHook applies the declared sampling policy during the Before stage.
// Sampled-out events throw SampledError; events without a policy pass through.
class SamplingHook(private val lookup: SamplingLookup) : UnimplementedHook() {
  override suspend fun before(hc: HookContext, hints: HookHints): EventEnvelope? {
    val policy = lookup(hc.eventName) ?: return null
    if (policy.strategy == SamplingStrategy.NONE) return null

    val keep =
        when (policy.strategy) {
          SamplingStrategy.USER_ID_HASH -> hashKeep(hc.context.userId, policy.rate)
          SamplingStrategy.RANDOM -> Math.random() < policy.rate
          SamplingStrategy.NONE -> true
        }
    if (!keep) throw SampledError()
    return null
  }
}

// fnv1a32 computes FNV-1a 32-bit hash, matching the Go and TypeScript implementations.
private fun fnv1a32(s: String): Long {
  val bytes = s.toByteArray(Charsets.UTF_8)
  var h = 2166136261L
  for (b in bytes) {
    h = h xor (b.toLong() and 0xFF)
    h = (h * 16777619L) and 0xFFFFFFFFL
  }
  return h
}

private fun hashKeep(userId: String, rate: Double): Boolean {
  val norm = fnv1a32(userId).toDouble() / 0x100000000L.toDouble()
  return norm < rate
}
